Code actions to configure severity and IDE code style options with .editorconfig

mavasani opened the pull request:
This PR revives the work done in https://github.com/dotnet/roslyn/pull/29332 to add a code action to configure severity of diagnostics and code action to configure code style option values. Unlike that PR, which relied on .editorconfig being marked as an additional file, this PR uses the newly added `AnalyzerConfigDocument` for .editorconfig files. Screenshots and a small working snippet is attached below:

**Top level menu:** Now we nest all configure and suppress code actions into a sub-menu to avoid clutter in lightbulb
![image](https://user-images.githubusercontent.com/10605811/58927565-fa917880-8703-11e9-9304-9e51510f4363.png)

**Create new .editorconfig with configured code style option value:**
![image](https://user-images.githubusercontent.com/10605811/58927869-282af180-8705-11e9-9d8d-e5ce1ad97902.png)

**Update existing .editorconfig with new code style option value:**
![image](https://user-images.githubusercontent.com/10605811/58927785-d3877680-8704-11e9-9f97-b5f3c78943ca.png)

**Create new .editorconfig with configured severity (also works for 3rd party analyzers):**
![image](https://user-images.githubusercontent.com/10605811/57649651-a8828a80-757d-11e9-8e35-6b338489807c.png)

**Update existing .editorconfig with new severity (also works for 3rd party analyzers):**
![image](https://user-images.githubusercontent.com/10605811/57649690-bb955a80-757d-11e9-8ff7-a73d5c0e9f25.png)

**Supports different syntax for IDE diagnostics with code style option(s):**
![image](https://user-images.githubusercontent.com/10605811/57650312-64908500-757f-11e9-8b1b-7b4d98d63853.png)

**Code action snippet:**
![Configure_Severity_Screenshot](https://user-images.githubusercontent.com/10605811/57649539-5fcad180-757d-11e9-8076-8d9ba615153b.gif)

**NOTE:** I also had to thread through AnalyzerConfigDocuments at bunch of places in the workspace layer to get the creating/updating of this new document kind via a code action to work as expected. I have added relevant tests.

**Future enhancements:**
Current code fix only adds/updates a single closest .editorconfig relative to the file with the diagnostic. In future, we might want to enable customization of this via options.

CyrusNajmabadi commented on src/EditorFeatures/Core.Wpf/Preview/PreviewFactoryService.cs:
```diff
@@ -494,6 +558,55 @@ public Task<object> CreateChangedAdditionalDocumentPreviewViewAsync(TextDocument
+            return CreateChangedDocumentViewAsync(
+                oldBuffer, newBuffer, description, originalLineSpans, changedLineSpans,
+                leftWorkspace, rightWorkSpace, zoomLevel, cancellationToken);
+        }
```
are these functions very duplicative with other functions that deal with normal/additional files?

mavasani replied:
Yes, almost everything I have added for AnalyzerConfigDocuments in this PR is clone of the functions for additional files. We can potentially share code by using individual document specific delegates as parameters of shared method or just keep the clones. I feel @jasonmalinowski is in the best position to make this call as he is the one who more often then not deals with the code in these area. I have taken the clone approach for now, and if required we can clean this up with follow-up PRs to share more code.

CyrusNajmabadi replied:
Ah yes, i believe i complained about the same thing with @jasonmalinowski about the same thing.  

For these sections we coudl at least try o have some helper shared methods...

jasonmalinowski replied:
For huge things like this I'd expect that we'd unify the code. If there's some small bits that can't quite make it unifiable like the add/remove or Open methods, my pattern was to unify to a helper that took lambdas that did those few bits. I think doing that also made it clear where the "differences" were and how we can then better unify code in the future.

mavasani replied:
Sounds good - I will update the PR to to share code for larger methods.

mavasani replied:
I ended up doing both the changes, as there seem to be lot of places where we pass the document kind as a boolean flag, and an enum is an obvious choice instead of multiple bools. I have kept the enum as internal for now, but we can make it public in future if we believe it is required and approved by the design team.

CyrusNajmabadi commented on src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:
```diff
@@ -484,33 +488,36 @@ private CodeRefactoring FilterOnUIThread(CodeRefactoring refactoring, Workspace
+                return;
+
+                // Local functions
+                SuggestedAction GetSuggestedAction(CodeAction action, CodeFix fix)
```
consider making static if possible.

mavasani replied:
That leads to 7-8 additional parameters, I'll keep it as is.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/CodeActions/CodeActionEditHandlerService.cs:
```diff
@@ -214,11 +214,13 @@ internal class CodeActionEditHandlerService : ForegroundThreadAffinitizedObject,
                 projectChange.GetRemovedAdditionalDocuments().Any() ||
                 projectChange.GetRemovedAnalyzerReferences().Any() ||
                 projectChange.GetRemovedDocuments().Any() ||
+                projectChange.GetRemovedAnalyzerConfigDocuments().Any() ||
```
any chance we could extract this out into separate PRs to make this easier to review?

mavasani replied:
I really wanted to extract these changes into a separate PR, but I wasn't able to verify the end-to-end functionality of code fix application, preview, opening these documents in editor, etc. without using the configure severity code action added in this PR. I can divide these into 2 PRs if this still seems critical for easier review.

CyrusNajmabadi replied:
It's 7k lines :)  Anythin to make it more manageable would be appreciated :D   Up to you though.  I can def get through this.  it'll just take a while.

CyrusNajmabadi commented on src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs:
```diff
@@ -499,19 +499,41 @@ private static async Task VerifyAgainstWorkspaceDefinitionAsync(string expectedT
+            return;
+
+            // Local functions.
+            void VerifyExpectedDocumentText(string expected, string actual)
```
static?

CyrusNajmabadi replied:
still applies.

CyrusNajmabadi commented on src/EditorFeatures/TestUtilities/CodeActions/AbstractCodeActionOrUserDiagnosticTest.cs:
```diff
@@ -499,19 +499,41 @@ private static async Task VerifyAgainstWorkspaceDefinitionAsync(string expectedT
+            {
+                if (expected == "")
+                {
+                    Assert.Equal((object)expected, actual);
```
i thought i wrote a comment why we have this behavior :)  can we move it over as well?

mavasani replied:
I did not find any comment in the original code, but I have tried to retain the original semantics.

CyrusNajmabadi commented:
Pausing on this to try to figure out a good way to order the review.

jasonmalinowski commented:
@mavasani Is this something we can break into multiple commits?

mavasani commented:
> Is this something we can break into multiple commits?

Let me break this into separate commits and force push...

mavasani commented:
@jasonmalinowski @CyrusNajmabadi I have updated the PR to be split into multiple commits, hopefully it is easier to review.

1) Commits 1-3: All the analyzer config document plumbing through Workspaces, EditorFeatures and Visual Studio layers
2) Commit 4: Analyzer config and additional document test infrastructure support
3) Rest of the commits - Support for the new code action and unit tests

I will now work on refactoring/sharing more code for analyzer config documents. Note that it is likely going to lead to helper methods with bunch of ugly lambdas, but probably that is better then cloning large pieces of code.

CyrusNajmabadi commented:
@mavasani Thanks for putting in that effort!

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Shared/Extensions/ISolutionExtensions.cs:
```diff
@@ -48,19 +49,22 @@ public static IEnumerable<DocumentId> GetChangedDocuments(this Solution newSolut
+            return solution.GetDocument(documentId) ?? solution.GetAdditionalDocument(documentId) ?? solution.GetAnalyzerConfigDocument(documentId);
         }
 
         public static Solution WithTextDocumentText(this Solution solution, DocumentId documentId, SourceText text, PreservationMode mode = PreservationMode.PreserveIdentity)
```
yikes.  this method def feels brittle.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Shared/Extensions/ISolutionExtensions.cs:
```diff
@@ -48,19 +49,22 @@ public static IEnumerable<DocumentId> GetChangedDocuments(this Solution newSolut
+                    return solution.WithAnalyzerConfigDocumentText(documentId, text, mode);
+
+                default:
+                    return solution.WithAdditionalDocumentText(documentId, text, mode);
```
can we check explicitly for AdditionalDocs, and then throw in the default case?  this seems like a bug waitin to happen in the future whne we add something new.  

ideally, we could just get rid of this entire method.  it seems fishy to me.

mavasani replied:
I am going to introduce a new sub-type of `TextDocument` for additional documents and handle them explicitly here and below method. I also think it would be good to introduce `TextDocumentKind` enum with 3 supported document kinds and expose it directly on `TextDocument` for clarity.

jasonmalinowski replied:
I'd be +1 for the sub-type since that was problematic for me in other cases too. Do we want an enum or just use pattern matching if you need a switch, now that we have that?

mavasani replied:
Sounds good - I'll hold off on the enum for now.

CyrusNajmabadi replied:
I'm fine with any approach.  I think Kinds+Subtypes fit together fine, allowing people to pick the pattern they think is most appropriate.  But i have no strong opinion here.  My main concern is just that we should not "defualt" to one doc type, but should strictly enumerate things and throw when we get something we don't expect.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Shared/Extensions/TextDocumentExtensions.cs:
```diff
@@ -12,12 +12,17 @@ internal static partial class TextDocumentExtensions
 
-            return textDocument.WithAdditionalDocumentText(text);
+                default:
+                    return textDocument.WithAdditionalDocumentText(text);
```
same comments apply here.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Workspace/AdhocWorkspace.cs:
```diff
@@ -201,5 +201,33 @@ public override void CloseAdditionalDocument(DocumentId documentId)
+                var text = doc.GetTextSynchronously(CancellationToken.None);
+                var version = doc.GetTextVersionSynchronously(CancellationToken.None);
+                var loader = TextLoader.From(TextAndVersion.Create(text, version, doc.FilePath));
+                this.OnAnalyzerConfigDocumentClosed(documentId, loader);
```
another +1 to @jasonmalinowski  on my feedback about encapsulting doc handling in some way so that each new doc type doesn't require this sort of copy/paste.  (i know you're aware of the feedback, but these prs just make me sad).

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Workspace/Solution/ProjectChanges.cs:
```diff
@@ -116,6 +116,17 @@ public IEnumerable<DocumentId> GetAddedAdditionalDocuments()
+                    yield return doc.Id;
+                }
+            }
+        }
```
nit: `_newProject.AnalyzerConfigDocuments.Where(d => !oldProject.Contains(d.Id)).Select(d => d.Id);`?

mavasani replied:
I have tried to retain the code style used in similar methods for existing document kinds :)

CyrusNajmabadi replied:
Ah, i missed that.  Yes.  if we're already doing the copy/paste approach, we should likely continue with that.  Only tweak the methods if you're then going to share between the 3 impls :)

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Workspace/Solution/ProjectChanges.cs:
```diff
@@ -191,5 +216,16 @@ public IEnumerable<DocumentId> GetRemovedAdditionalDocuments()
+                    yield return doc.Id;
+                }
+            }
+        }
```
@jasonmalinowski How opposed would you be to a PR that tried to unify all this duplication?  i'm potentially interested in writing it, but i'd prefer to only do it if i had reasonable confidence it would be taken.

mavasani replied:
I am +1 on a follow-up PR to encapsulate all the document handling in a central way, without us having to handle all document kinds at every place in different layers, and then have helpers with bunch of lambdas to share code. I will still try to do the later for large methods that I have cloned in this PR, but still like the overall idea. It would be super beneficial if we add a new document kind in future.

jasonmalinowski replied:
Yes, I'm all for cleaning this up, at least in principle. I think a lot of this can be done just by extracting out of ProjectState a "abstract collection of documents" thingy of some sort that has lots of these common operations, and so then we're working with those internally. I'm not sure if that's something we'd expose or still just have the (current) public methods just forward to it; I'm actually OK with that being internal implementation since the forwarding methods would be trivial.

If you have a concrete pattern in your head it might be worth discussing prior doing it so we can try to forsee gotchas, but I fully confess if it was me trying to do the dedup I'd just dive in head first because you'll learn things much faster that way.

CyrusNajmabadi replied:
That's pretty much exactly waht i was thinking.  On our public surface area, we'd still expose the split.  But all those public methods would just callinto the raw instance of the specific collection which had all the shared logic.  So you'd basically have:

```C#
class Solution {
    _normalDocuments = new NormalDocumentCollection();
    _analyzerDocuments = new AnalyzerDocumentCollection();
    _configDocuments = new ConfigDocumentCollectoin();
}

class AbstractDocumentCollection<TDocument> where TDocument : TextDocument {
    // all our algorithms and query/mutation methods.
    abstract void Something();
}

class ConfigDocumentCollection : AbstractDocumentCollection<ConfigDocument> {
    // overrides to make tings specific to config docs.
}
```

The leaf DocCollections would be extremely simple, and we'd put all shareable logic in one place.

CyrusNajmabadi commented:
Done with 4c0459a25a556db891b467ed713e53415374095b.  Small nits, some concerns about default switch behavior.  Strongly think we should invest in a non-duplicative approach to files.  But that's a non-blocking concern.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Preview/FileChange.cs:
```diff
@@ -210,7 +210,8 @@ public TextDocument GetUpdatedDocument()
 
-        public bool IsAdditionalDocumentChange => !((_left ?? _right) is Document);
+        public bool IsAdditionalDocumentChange => !((_left ?? _right) is Document) && !IsAnalyzerConfigDocumentChange;
+        public bool IsAnalyzerConfigDocumentChange => (_left ?? _right) is AnalyzerConfigDocument;
```
ick, i don't like this implicit relationship.  basically, if something is an AdditionalDoc i'd like that to be explicit.  i don't want it to just fall out because it's "not an analyzer doc".  the latter form just means more complexity pushed to the edges if/when we ever add new doc types.

mavasani replied:
Should be cleaner now with document kind check.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Preview/PreviewUpdater.PreviewDialogWorkspace.cs:
```diff
@@ -25,25 +25,37 @@ public void UnregisterTextContainer(SourceTextContainer container)
+                        break;
+
+                    default:
+                        OnAdditionalDocumentClosed(document.Id, new PreviewTextLoader(text));
```
same comment from previous PR about defaults and additional docs.  personally, i don't even like what we had before.  but at least it was only "real doc + additional doc".  now that we're adding more types, having this sort of 'implicit additional default' feels very wrong to me.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Preview/PreviewUpdater.PreviewDialogWorkspace.cs:
```diff
@@ -25,25 +25,37 @@ public void UnregisterTextContainer(SourceTextContainer container)
+
+                    default:
+                        OpenAdditionalDocument(document.Id);
+                        break;
```
same comment as above.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Preview/TopLevelChange.cs:
```diff
@@ -129,6 +131,8 @@ private Solution ApplyFileChanges(Solution solution, IEnumerable<FileChange> fil
                             solution.RemoveAdditionalDocument(updatedDocumentIdOpt) :
+                            isAnalyzerConfigDoc ?
+                            solution.RemoveAnalyzerConfigDocument(updatedDocumentIdOpt) :
                             solution.RemoveDocument(updatedDocumentIdOpt);
```
nit: switch to `?:` on subsequent lines to better match IDE coding patterns (and make this easier to read :)).

nit: these sorts of fall-through patterns concern me (As per previous comments).

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/Preview/TopLevelChange.cs:
```diff
@@ -95,15 +95,15 @@ private Solution ApplyFileChanges(Solution solution, IEnumerable<FileChange> fil
                     {
                         ApplyFileChangesCore(oldTextDocument, newLinkedDocumentId, updatedDocumentTextOpt,
-                            fileChange.CheckState, fileChange.IsAdditionalDocumentChange);
+                            fileChange.CheckState, fileChange.IsAdditionalDocumentChange, fileChange.IsAnalyzerConfigDocumentChange);
```
I feel that there should be an enum stating the type of the doc (Normal/Additional/Analyzer), and we can pass that along.  code that processed can operate on that enum and throw when they encounter a new type.  that will make our code at least *throwing* when we add new types, instead of silently doing the wrong thing.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.AddAnalyzerConfigDocumentUndoUnit.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    internal partial class VisualStudioWorkspaceImpl
+    {
+        private class AddAnalyzerConfigDocumentUndoUnit : AbstractAddDocumentUndoUnit
```
i am very happy we could share code here :)

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs:
```diff
@@ -712,16 +718,24 @@ private OleInterop.IOleUndoManager TryGetUndoManager()
         protected override void ApplyDocumentAdded(DocumentInfo info, SourceText text)
         {
-            AddDocumentCore(info, text, isAdditionalDocument: false);
+            AddDocumentCore(info, text, isAdditionalDocument: false, isAnalyzerConfigDocument: false);
```
same comment about this benefitting from a distinct enum.

CyrusNajmabadi commented on src/VisualStudio/Core/Impl/RoslynVisualStudioWorkspace.cs:
```diff
@@ -72,7 +72,7 @@ internal override IInvisibleEditor OpenInvisibleEditor(DocumentId documentId)
             }
 
-            var document = this.CurrentSolution.GetDocument(documentId) ?? this.CurrentSolution.GetAdditionalDocument(documentId);
+            var document = this.CurrentSolution.GetTextDocument(documentId);
```
ok, based on this, i see hte value in just GetTextDocument.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpDiagnosticIdToEditorConfigOptionMappingService.cs:
```diff
@@ -0,0 +1,93 @@
+                    return (optionOpt: null, handled: false);
+            }
+
+            return (optionOpt: option, handled: true);
```
this is *super* ick... I'm very very opposed to the idea that we add new diagnostic features, but then have to update a dozen or so independent parts of the IDE codebase in order for them to "light up" all these other features properly.  I am strongly pushing that we stop continually doing this and we go back and reassess how DiagnosticAnalyzers work in order to have them properly enhance the IDE experience.

For example, perhaps there can be a new interface that a DiagnosticAnalyzer optionally implements (or no-op virtual methods they override).  These can supply this additional information that these auxilliary features need.

mavasani replied:
This is an interim solution, I completely agree we need a better solution, in fact something that works even for third party analyzers who want to be configurable through .editorconfig code style options. However, that is a much bigger work item. I have still ensured that for the current approach we have unit tests that ensure that every new IDE diagnosticID is handled explicitly here.  If someone adds a new IDE diagnostic ID, added unit test will fail with a message guiding how to handle it correctly.

> perhaps there can be a new interface that a DiagnosticAnalyzer optionally implements 

This approach does not help here. We need a generic mapping from a reported diagnostic/diagnosticID to corresponding code style option used to configure it, so that any configuration code action can tweak this option in analyzer config document.

CyrusNajmabadi commented on src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs:
```diff
@@ -651,17 +669,23 @@ private static ImmutableArray<string> GetAndTestFixableDiagnosticIds(CodeFixProv
+                    builder.Add(languageKindAndFixersValue.Value);
+                }
+
+                var lazySuppressionOrConfigurationFixers = new Lazy<ImmutableArray<ISuppressionOrConfigurationFixProvider>>(() => builder.ToImmutableAndFree());
```
i don't get how this Lazy helps out here.  You've done all the work *prior* to instantiating it.  so might as well just pass along the array.  If this really does need to be lazy, theabove work should move into the lambda.

CyrusNajmabadi commented on src/EditorFeatures/TestUtilities/Diagnostics/AbstractUserDiagnosticTest.cs:
```diff
@@ -195,10 +195,6 @@ protected Document GetDocumentAndAnnotatedSpan(TestWorkspace workspace, out stri
 
-            actions = actions.SelectMany(a => a is TopLevelSuppressionCodeAction
-                ? a.NestedCodeActions
-                : ImmutableArray.Create(a)).ToImmutableArray();
```
why removed?

mavasani replied:
The logic was specific to suppression actions, so has been moved into the MassageActions override in AbstractSuppressionDiagnosticTest

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/AbstractTopLevelConfigurationOrSuppressionCodeAction.cs:
```diff
@@ -0,0 +1,18 @@
+        }
+
+        // Put configurations/suppressions at the end of everything.
+        internal override CodeActionPriority Priority => CodeActionPriority.None;
```
interesting.  is that where 'None' goes?  it's lower than `.Low`?  Or does this mean: the  codefix infrastructure will determine where this goes.

mavasani replied:
The suppression code fix already had None priority [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.Features/CodeFixes/Suppression/TopLevelSuppressionCodeAction.cs,20), and the ordering logic is already ensuring that these go at the very bottom of the lightbulb.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.ConfigurationUpdater.cs:
```diff
@@ -0,0 +1,357 @@
+
+            // Regular expression for .editorconfig header.
+            // For example: "[*.cs]    # Optional comment"
+            private static readonly Regex s_headerPattern = new Regex(@"\[\*([^#\[]*)\]([ ]*[;#].*)?");
```
are these copied from the code in the compiler that does similar matching (that @agocke wrote)?  Can we possibly have a linked file with these regexes to share them?

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.ConfigurationUpdater.cs:
```diff
@@ -0,0 +1,357 @@
+
+            // Regular expression for .editorconfig code style option entry.
+            // For example: "dotnet_style_object_initializer = true:suggestion   # Optional comment"
+            private static readonly Regex s_optionBasedEntryPattern = new Regex(@"([\w ]+)=([\w ]+):[ ]*([\w]+)([ ]*[;#].*)?");
```
should these regexes have `start/end of line` chars in them so that they don't match in the middle of some line?  @agocke for thoughts on this as well

mavasani replied:
The parsing logic that uses this regex reads the editorconfig file contents one line at a time, so the behavior you mentioned is already ensured.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.ConfigurationUpdater.cs:
```diff
@@ -0,0 +1,357 @@
+
+            // Regular expression for .editorconfig diagnosticID severity configuration entry.
+            // For example: "dotnet_diagnostic.CA2000.severity = suggestion   # Optional comment"
+            private static readonly Regex s_severityBasedEntryPattern = new Regex(@"([\w \.]+)=[ ]*([\w]+)([ ]*[;#].*)?");
```
note:  i haven't actually looked at the contents of the regex yet.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.ConfigurationUpdater.cs:
```diff
@@ -0,0 +1,357 @@
+                CancellationToken cancellationToken)
+            {
+                var updater = new ConfigurationUpdater(severity, diagnostic, project, cancellationToken);
+                return await updater.Configure().ConfigureAwait(false);
```
async/await can all be removed from this static helper method.

mavasani replied:
Done. A code fix to detect and auto-fix this would be nice :-)

CyrusNajmabadi replied:
i agree.  though there's a slight and subtle change in semantics that can occasionally be impactful.  It's not an issue here... but it's something we might get dinged on if it impacts someone negatively. :)

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.ConfigurationUpdater.cs:
```diff
@@ -0,0 +1,357 @@
+                var solution = _project.Solution;
+
+                // Find existing .editorconfig or generate a new one if none exists.
+                var editorConfigDocument = FindOrGenerateEditorConfig(ref solution);
```
returning doc, and passin solutoin by ref is *super* wonky.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.ConfigurationUpdater.cs:
```diff
@@ -0,0 +1,357 @@
+            {
+                if (_project.AnalyzerConfigDocuments.Any())
+                {
+                    var diagnosticFilePath = PathUtilities.GetDirectoryName(_diagnostic.Location.SourceTree?.FilePath ?? _project.FilePath);
```
these might both be null.  Is GetDirectoryName and IsAbsolute resilient to that?

mavasani replied:
Yes.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.ConfigurationUpdater.cs:
```diff
@@ -0,0 +1,357 @@
+                    {
+                        var analyzerConfigDirectory = PathUtilities.GetDirectoryName(analyzerConfigDocument.FilePath);
+                        if (diagnosticFilePath.StartsWith(analyzerConfigDirectory) &&
+                            analyzerConfigDirectory.Length > bestPath.Length)
```
i personally find this check confusing and can't convince myself it is right.  for example, say you have  `foo/a/b/c/d/e` and `foo/laskdjlksajdlakjsldkjsadlkjasldjasldjaskdjas`, the latter would preferable right?  but it's longer than the former.

mavasani replied:
`diagnosticFilePath.StartsWith(analyzerConfigDirectory)` check ensures that we only consider ancestor directories of `diagnosticFilePath`. Only one of the above two paths in your example would be a valid candidated. Additionally, this is just a simple heuristic, and I fully anticipate we will have to fine tune it in future.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.ConfigurationUpdater.cs:
```diff
@@ -0,0 +1,357 @@
+                var editorConfigOption = diagnosticIdToEditorConfigMappingService.GetMappedEditorConfigOption(_diagnostic.Id);
+                if (editorConfigOption != null)
+                {
+                    var editorConfigLocation = editorConfigOption.StorageLocations.OfType<IEditorConfigStorageLocation2>().FirstOrDefault();
```
we have an IEditorConfigStorageLocation2??

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.ConfigurationUpdater.cs:
```diff
@@ -0,0 +1,357 @@
+                            var match = s_optionBasedEntryPattern.Match(editorConfigString);
+                            if (match.Success)
+                            {
+                                return (optionName: match.Groups[1].Value.Trim(), optionValue: match.Groups[2].Value.Trim());
```
this is why i far prefer named-groups in a regex.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.ConfigurationUpdater.cs:
```diff
@@ -0,0 +1,357 @@
+                }
+
+                return null;
+            }
```
skipped this method.

mavasani replied:
Note that you already reviewed majority of this code last year in the original PR https://github.com/dotnet/roslyn/pull/29332. I have tried to preserve most of the logic from that PR in the logic to add/update editorconfig entry.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.ConfigurationUpdater.cs:
```diff
@@ -0,0 +1,357 @@
+                else if (_language == LanguageNames.VisualBasic && headers.Any(header => header.Key.Contains(".vb")))
+                {
+                    existingHeaderOpt = headers.FirstOrDefault(header => header.Key.Contains(".vb")).Value;
+                }
```
nit: consider ternary expression.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.ConfigurationUpdater.cs:
```diff
@@ -0,0 +1,357 @@
+                }
+
+                return null;
+            }
```
skipped tihs method as well.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.cs:
```diff
@@ -0,0 +1,69 @@
+    internal sealed partial class ConfigureSeverityLevelCodeFixProvider : ISuppressionOrConfigurationFixProvider
+    {
+        public bool IsFixableDiagnostic(Diagnostic diagnostic)
+            => !diagnostic.IsSuppressed && !SuppressionHelpers.IsNotConfigurableDiagnostic(diagnostic);
```
can this be commented?  for example, i don't get why you couldn't configure the severity if the diagnostic is suppressed...

mavasani replied:
Actually, I don't think the code fix engine ever attempts to find fixes for suppressed diagnostics. I have just added a defensive check. I will add comment.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.cs:
```diff
@@ -0,0 +1,69 @@
+                        solution => ConfigurationUpdater.ConfigureEditorConfig(EditorConfigSeverityStrings.Warning, diagnostic, project, cancellationToken)));
+                nestedActions.Add(
+                    new SolutionChangeAction(
+                        "Error",
```
can use 'nameof' for all these raw string literals.

CyrusNajmabadi replied:
could have an array/list of all those strings, and then loop.

CyrusNajmabadi commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigureSeverityLevelCodeFixProvider.cs:
```diff
@@ -0,0 +1,69 @@
+                        "Error",
+                        solution => ConfigurationUpdater.ConfigureEditorConfig(EditorConfigSeverityStrings.Error, diagnostic, project, cancellationToken)));
+
+                var codeAction = new TopLevelConfigureSeverityCodeAction(diagnostic, nestedActions.ToImmutable());
```
1. arraybuilders are cheap.  much cleaner to just put hte builder inside the loop, add to it, and do ToImmutableAndFree here.
2. are these items only created when the list is expanded?  or are they created for all fixes across the solution.  If the former, we don't really need to hyperoptimize like this.

CyrusNajmabadi commented on src/EditorFeatures/Core.Wpf/EditorFeaturesWpfResources.resx:
```diff
@@ -153,4 +153,7 @@
     <value>Regex - Other Escape</value>
   </data>
+  <data name="Configure_or_Suppress_issues" xml:space="preserve">
+    <value>Configure or Suppress issues</value>
```
casing is wonky.  Should be `Configure or suppress issues`.

mavasani replied:
I have explicitly kept "Suppress" with capital S, given that we used to almost always have a top level item for Suppress that is now nested. I can change if you/others strongly feel otherwise.

CyrusNajmabadi commented on src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:
```diff
@@ -476,6 +477,9 @@ private CodeRefactoring FilterOnUIThread(CodeRefactoring refactoring, Workspace
             }
 
+            private static bool IsTopLevelSuppressionAction(CodeAction action)
+                => action is AbstractTopLevelConfigurationOrSuppressionCodeAction;
```
we have AbstractTopLevelConfigurationOrSuppressionCodeAction.  does it need to be abstract?  what are the subclasses of it?  Feels like it could just be the TopLevelConfigurationAction, and the supression/severity items are just nested in that.

mavasani replied:
`AbstractTopLevelConfigurationOrSuppressionCodeAction` represents the top level code action created by each configuration/suppression provider. Then the engine groups them together and wraps all these per-provider top level actions under another topmost level suggested action for that shows up in the light bulb menu.

CyrusNajmabadi commented on src/EditorFeatures/Core.Wpf/Suggestions/SuggestedActionsSource.cs:
```diff
@@ -582,27 +589,100 @@ private CodeRefactoring FilterOnUIThread(CodeRefactoring refactoring, Workspace
+            {
+                int minStart = -1;
+                int maxEnd = -1;
+                string category = PredefinedSuggestedActionCategoryNames.CodeFix;
```
can we add some comments explaining what's going on.

mavasani replied:
Added comments.

CyrusNajmabadi commented:
Overall this is very promising.  I 100% want this feature for customers.  And i think this PR is overall a great way to get us there.  I do worry about what i see as continued approaches to expediently getting us features that continue adding costs and taxes to the codebase.  It's usually the cheaper way to get the feature out more quickly.  But i feel it ends up costing us more in the long term.

CyrusNajmabadi commented:
Note: i would break this into two PRs.  The non-feature work (i.e. all the workspaces stuff) then the feature work.  The former looks very close to being done.  We could likely take as is, or (my preference) have some small design tweaks made to it to be a little more untrusting of the data that's flying around :)

For the feature work, i really do think we should consider certain parts of the design and think about how to properly design the internal APIs to support it in a less 'hacky' manner.  

Thanks much!

mavasani commented:
@jasonmalinowski @CyrusNajmabadi I have pushed few more commits to address some feedback:

1. https://github.com/dotnet/roslyn/pull/35691/commits/5f54086de7e72d1bd6d5b45c90ccf5dfdb1948f7 and https://github.com/dotnet/roslyn/pull/35691/commits/4fff727560099979a6fb24bf49b7b2118d0311ab: Share more code between documents, additional documents and analyzer config documents handlers.

2. https://github.com/dotnet/roslyn/pull/35691/commits/bae75b4972847d6284b7e61ca9631a690787fc55: Add `AdditionalDocument : TextDocument` sub-type and `TextDocumentKind` enum and change the existing methods switching on explicit TextDocument sub-types and those using one or more bools for document kind to now just work with the TextDocumentKind.

CyrusNajmabadi commented:
So i'm basically nearly entirely on board with the 'normal/analyzer/additional docs' change.  If it could be extracted, there would essentially only be optional nits, and i'd be very 👍 on getting it in with any future changes up to you.

tying this to the feature makes this less palatable for me in the current state due to me wanting to really questoin and discuss how best to get the the feature level code in.

So it's up to you on how to proceed.  This is all optional feedback.  And i get the line you guys need to cross between delivering value versus dealing with debt.  So thnk about it and decide what you think is best overall.  I won't mind any direction you take as long as it's an informed and understood choice. 

Thanks!

CyrusNajmabadi commented:
> Yes, sure. (Regular) Documents have some special handling in some of these handlers, while additional document and analyzer config documents are similar due to them being plain text documents. I will add a comment

perfect.  taht's what i figured.  but it would still be good to have the comment to understand.  at the very least, it will help us understand what does/doesn't work in preview for these 'plain text' docs.

mavasani commented:
> tying this to the feature makes this less palatable for me in the current state due to me wanting to really questoin and discuss how best to get the the feature level code in.

I am fine with separating this into 2 separate PRs if @jasonmalinowski and other folks from @dotnet/roslyn-ide feel the same. The only concern would be that the added configuration feature is the only way to test/evaluate the analyzer config document related code changes in the Visual Studio and EditorFeatures layers. If we are doing a big refactoring/updates on the added code, we wouldn't have a way to test them even manually without the feature. This is not a big concern if the second PR can go in soon-ish as well.

mavasani commented:
@jasonmalinowski @CyrusNajmabadi - I have created a separate [PR](https://github.com/dotnet/roslyn/pull/35762) cherry-picking the AnalyzerConfigDocument related changes from here.

mavasani commented:
Marked this PR as blocked on https://github.com/dotnet/roslyn/pull/35762. I will repurpose this PR for the configure severity code action support once https://github.com/dotnet/roslyn/pull/35762 has been merged in.

mavasani commented:
Core analyzer config changes have been merged (#35762) and I have force pushed so this PR only has commits relevant to the new Configure severity code action. I will address the issues raised by @CyrusNajmabadi and push a follow-up commit. Thanks!

mavasani commented:
> 9a0e198. Strong opposition. I think this is a cutting-corners approach to this problem to avoid a proper design that any analyzer could play in. I've given similar feedback in the past about how our features which should be isolated are now leaking all over the codebase to make other features work. This just continues that and i think it needs to stop.

@CyrusNajmabadi - I have reverted https://github.com/dotnet/roslyn/pull/35691/commits/1fb9371c4c88b47b4e004e78e2d820147fe2336c for tracking mapping of diagnostic IDs to code style options. I have taken an alternate approach in https://github.com/dotnet/roslyn/pull/35691/commits/5b6cee9c53a2a8dbf4bed8d473e77fd740dff114, which now requires each analyzer to explicitly provide the associated option with each supported descriptor. I will look at your other review comments now.

CyrusNajmabadi commented:
> @CyrusNajmabadi - I have reverted 1fb9371 for tracking mapping of diagnostic IDs to code style options. I have taken an alternate approach in 5b6cee9, which now requires each analyzer to explicitly provide the associated option with each supported descriptor. I will look at your other review comments now.

@mavasani Thanks for the change!  I'm much happier with it.  It def satisfies at least one of my goals, which is that the beahvior and description of the feature is contained within the feature itself.  My other goal (which is absolutely not required and can come later) would be that this be a general part of the code-fixer infrastructure, and not limited to just our internal features.  

My objection is withdrawn on that commit now.  Thanks!

mavasani commented:
@CyrusNajmabadi I have addressed most of your review feedback with the latest commit.

@dotnet/roslyn-ide @heejaechang - would you be able to review this PR?

CyrusNajmabadi commented:
> @CyrusNajmabadi I have addressed most of your review feedback with the latest commit.

@mavasani Thanks!  Will try to review this this weekend.  I don't expect any issues, your responses all seemed sensible to me :)

mavasani commented:
FYI: I have pushed a new commit https://github.com/dotnet/roslyn/pull/35691/commits/f1511b887bd36e89a167eb959eed6de4d20ca454, which adds a configuration fixer to enable configuring IDE code style option values from the lightbulb. I have updated the PR description and screenshots.

Tagging @jinujoseph @kendrahavens @mikadumont

CyrusNajmabadi commented:
the pictures are awesome.  Can you show what it looks like to change the severity in this case:

![image](https://user-images.githubusercontent.com/4564579/58928468-6cb78c80-8707-11e9-803d-7584daa29402.png)

mavasani commented:
![image](https://user-images.githubusercontent.com/10605811/58928747-ab017b80-8708-11e9-9a6f-f1e14e1a7552.png)

I currently chose the implementation to add/update all code style option based entries instead of using `dotnet_diagnostic.IDE0008.severity = error`, which might add confusion as a subsequent entry for code style option will conflict with this entry. I felt it would be good to stick to following principle:

1. Always use code style option based syntax (i.e. `optionName = optionValue:severity`) for IDE code style diagnostics.
2. Always use compiler based syntax (i.e. `dotnet_diagnostic.ID.severity = severity`) for 3rd party diagnostics and non-code style IDE diagnostics.

mavasani commented on src/Features/Core/Portable/CodeFixes/Configuration/ConfigurationUpdater.cs:
```diff
@@ -0,0 +1,506 @@
+                    {
+                        var allHeaders = mostRecentHeader.Split(',', '.', ' ', '{', '}');
+                        if ((_language == LanguageNames.CSharp && !allHeaders.Contains("cs")) ||
+                            (_language == LanguageNames.VisualBasic && !allHeaders.Contains("vb")))
```
Yes, you are correct. In fact, there are lot of enhancements that I have in mind here - let me file a separate tracking issue for all the future improvements.

mavasani replied:
Filed https://github.com/dotnet/roslyn/issues/36348
